### PR TITLE
feat(log): export log to clickhouse

### PR DESCRIPTION
**Pull request by hunshenshi**
1. add endpoint in config.yml, like this `SRV_APPLET_MGR__Logger_CKEndpoint: "clickhouse://127.0.0.1:9000/w3b"`
2. create db in the clickhouse use `create database w3b`
3. create table in the `w3b` db use 
`CREATE TABLE IF NOT EXISTS server_logs (
     Timestamp DateTime64(9) CODEC(Delta, ZSTD(1)),
     ProjectName String CODEC(ZSTD(1)),
     Msg String CODEC(ZSTD(1))
) ENGINE MergeTree()
PARTITION BY toDate(Timestamp)
ORDER BY (toUnixTimestamp(Timestamp), ProjectName)
SETTINGS index_granularity=8192, ttl_only_drop_parts = 1;`

### CODE REVIEW

**Comment by Liuhaai** [comment]
could the client of [Clickhouse](https://github.com/machinefi/w3bstream/blob/main/pkg/modules/metrics/clickhouse.go) be shared?

**Comment by Liuhaai** [comment]
https://clickhouse.com/blog/storing-log-data-in-clickhouse-fluent-bit-vector-open-telemetry#agents

File: pkg/depends/conf/log/log.go [hunk 1]
```diff
@@ -69,6 +80,106 @@ func (l *Log) Init() {
...
+	conn, err := sql.Open("clickhouse", ck.CKEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+func (ck *ClickhouseHook) ping() bool {
+	if err := ck.ckDB.Ping(); err != nil {
+		return false
```
Comment (CoderZhi) [inline pkg/depends/conf/log/log.go:141]: print err in debug level

File: pkg/depends/conf/log/log.go [hunk 2]
```diff
@@ -69,6 +80,106 @@ func (l *Log) Init() {
...
+	return err
+}
+
+func (ck *ClickhouseHook) doWithTx(fn func(tx *sql.Tx) error) error {
+	tx, err := ck.ckDB.Begin()
+	if err != nil {
+		return fmt.Errorf("ck tx begin error: %w", err)
+	}
+	defer func() {
+		_ = tx.Rollback()
```
Comment (CoderZhi) [inline pkg/depends/conf/log/log.go:175]: print err if any

File: pkg/depends/conf/log/log.go [hunk 3]
```diff
@@ -69,6 +80,106 @@ func (l *Log) Init() {
...
+}
+
+func (ck *ClickhouseHook) insertLogs(entries []logrus.Entry) error {
+	err := ck.doWithTx(func(tx *sql.Tx) error {
+		statement, err := tx.PrepareContext(context.Background(), ck.insertSql)
+		if err != nil {
+			return fmt.Errorf(" ck prepareContext error:%w", err)
+		}
+		defer func() {
+			_ = statement.Close()
```
Comment (CoderZhi) [inline pkg/depends/conf/log/log.go:153]: print err in some way

File: pkg/depends/conf/log/log.go [hunk 4]
```diff
@@ -52,6 +57,12 @@ func (l *Log) InitLogrus() {
 	// TODO add hook with goid meta logrus.AddHook(goid.Default)
 	logrus.AddHook(&ProjectAndMetaHook{l.Name})
 
+	if l.CKEndpoint != "" {
+		ckHook, err := newClickhouseHook(l.CKEndpoint)
+		if err == nil {
+			logrus.AddHook(ckHook)
```
Comment (Liuhaai) [inline pkg/depends/conf/log/log.go:63]: fire connecting for each log?

File: pkg/depends/conf/log/log.go [hunk 5]
```diff
@@ -69,6 +80,106 @@ func (l *Log) Init() {
...
+	CKEndpoint     string
+	ckDB           *sql.DB
+	insertSql      string
+	entries        []logrus.Entry
+	lastInsertTime time.Time
+	duration       time.Duration
+	lock           *sync.Mutex
+}
+
+func newClickhouseHook(endpoint string) (*ClickhouseHook, error) {
```
Comment (Liuhaai) [inline pkg/depends/conf/log/log.go:95]: could this client be shared with https://github.com/machinefi/w3bstream/blob/53ba419b3f6eb424e61ce3fbb1b251c4a762a1df/pkg/modules/metrics/clickhouse.go

**Comment by Liuhaai** [comment]
IMO, it would be easier to ship the log file from docker with `filebeat`. No dependencies on the backend code.

File: pkg/depends/conf/log/log.go (hunk 5 above)
Comment (hunshenshi) [reply pkg/depends/conf/log/log.go:95]: The log should be a independent endpoint, so they are no dependency.

**Comment by hunshenshi** [comment]
>

**Comment by hunshenshi** [comment]
> IMO, it would be easier to ship the log file from docker with `filebeat` (or simiar services) for ELK stack. No dependencies on the backend code.

yes, but they require a separate component to implement,  and they are not friendly to our  developer.

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**